### DISCUSSION TITLE
Calendar event enumerator

### DISCUSSION THREAD

**Discussion by cosinekitty** [cosinekitty/astronomy#110] [General]
@matheo I have started a prototype of the calendar event enumerator as we discussed earlier. It is in the [calendar2](https://github.com/cosinekitty/astronomy/tree/calendar2) branch. Take a look at [calendar.ts](https://github.com/cosinekitty/astronomy/blob/calendar2/demo/nodejs/calendar/calendar.ts). 

The idea is that there are event enumerator classes you can create that follow a simple FindFirst/FindNext interface. Then you feed them into an `EventCollator` which merges the multiple event streams into one chronological stream. It does the minimum amount of work to iterate one event at a time, and uses only a small amount of memory. I think this would be a pretty good algorithm to feed a dynamic calendar inside a UI.

The example output is in [calendar_correct.txt](https://github.com/cosinekitty/astronomy/blob/calendar2/demo/nodejs/calendar/test/calendar_correct.txt), which is a flat listing of sunrise, sunset, moonrise, moonset, season changes (equinoxes and solstices), and moon phases. I will add many more enumerators of interesting events soon (eclipses, conjunctions, etc).

I'll be interested to know what you think. Is this something you could build a calendar UI from?

**Comment by matheo** [comment]
Wooow yess!
seems enough data to build a nice UI :)
awesome stuff! thank you!

**Comment by cosinekitty** [comment]
OK, great! I will add some more enumerators in the calendar2 branch over the next few days. Basically I will add anything I can think of that I would personally want to see on an astronomy calendar.

**Comment by cosinekitty** [comment]
Tonight I added the following calendar enumerators:

- planetary conjunctions (when planets are invisible because of appearing too close to the Sun)
- planetary oppositions (when outer planets are closest to the Earth and best visible)
- max elongations of Mercury and Venus (when they appear at the largest angle from the Sun)
- Venus peak apparent magnitude (when Venus looks the brightest)

The remaining ideas on my list to be completed are:

- Lunar and solar eclipses
- Transits of Mercury and Venus (when they pass between the Sun and Earth)
- Lunar apogee and perigee (moon farthest and closest to Earth)
- Planet aphelion and perihelion (planet farthest and closest to the Sun)
- When planets enter a new constellation
- Moon and Sun culmination (time of highest angle above the horizon every day)

**Reply by matheo** [reply]
You can't imagine how badly I'd like to have your astro+math skills 😎
Thanks again for sharing them!

**Reply by cosinekitty** [reply]
Thanks, @matheo ! 

Tonight I finished lunar eclipses, solar eclipses, and transits of Mercury and Venus.  I will continue working through the list I wrote above.

**Comment by cosinekitty** [comment]
I finished the calendar demo and it is now merged into the master branch. Its output is quite lengthy. For a real calendar, I suspect we would omit sunrise, sunset, moonrise, moonset, and Sun & Moon culmination events. If someone clicked on a particular calendar day, we could show all of those things, but for the overall calendar view, it would be a lot more efficient to only calculate events that don't happen every single day.

**Reply by matheo** [reply]
> it would be a lot more efficient to only calculate events that don't happen every single day

I completely agree
